File: src/json_db.py
from dataclasses import dataclass
import json
import os

@dataclass
class Item:
    id: str
    title: str
    description: str
    price: int #price in cents

class JSON_DB:
    FILEPATH = None
    # Keeps copy of data locally for faster reading
    # This is updated first, and then written to file on any change
    DATA = {
        "items": {
            "item-id": {
                "title": "test item",
                "description": "this is a test",
                "price": 34_01,
                "images": ["testitem.png"]
            }
        }
    }

    @staticmethod
    def save_data():
        with open(JSON_DB.FILEPATH, "w") as f:
            f.write(json.dumps(JSON_DB.DATA))

    @staticmethod
    def connect(filepath):
        # Bugs out when file exists but doesn't have a valid data structure
        JSON_DB.FILEPATH = filepath
        if os.path.exists(JSON_DB.FILEPATH):
            with open(JSON_DB.FILEPATH, "r") as f:
                JSON_DB.DATA = json.loads(f.read())
        else:
            JSON_DB.save_data()

    @staticmethod
    def add_item(item: Item):
        JSON_DB.DATA["items"][item.id] = {"title": item.title, "description": item.description, "price": item.price}
        JSON_DB.save_data()

    @staticmethod
    def delete_item(item_id):
        if not item_id in JSON_DB.DATA["items"]:
            return
        del JSON_DB.DATA["items"][item_id]
        JSON_DB.save_data()

    @staticmethod
    def get_items():
        return JSON_DB.DATA

File: src/json_db.js
const { files } = require("./util")

const json_db = (() => {

    let db_data = { items: {}, store_text: "" }
    let db_path = ""

    async function save_db_data() {
        await files.write(db_path, JSON.stringify(db_data))
    }

    return {
        connect: async (filepath) => {
            const file_exists = await files.exists(filepath)
            db_path = filepath
            if (file_exists) {
                const data = await files.read(db_path)
                if (data) {
                    db_data = JSON.parse(data)
                } else {
                    console.error(`[DATABASE ERROR] Database file was read but no data was received, check integrity of file ${db_path}`)
                }
            } else {
                save_db_data()
            }
        },
        get_data: () => {
            return db_data
        },
        add_item: async (id, title, description, price, reserved, images) => {
            db_data.items[id] = {title, description, price, reserved, images}
            save_db_data()
        },
        delete_item: async (item_id) => {
            if (item_id in db_data.items) {
                delete db_data.items[item_id]
                save_db_data()
            }
        },
        set_reserved: async (item_id, reserved) => {
            if (!item_id in db_data.items) {
                return
            }
            db_data.items[item_id]["reserved"] = reserved
            save_db_data()
        },
        is_reserved: (item_id) => {
            if (!item_id in db_data.items) {
                return
            }
            return db_data.items[item_id]["reserved"]
        },
        set_store_text: (text) => {
            if (!typeof text == "string") return
            db_data.store_text = text
            save_db_data()
        },
        get_store_text: () => {
            return db_data.store_text
        }
    }
})()

module.exports = json_db